### PR TITLE
[VarDumper] Use \ReflectionReference for determining if a key is a reference (php >= 7.4)

### PR DESCRIPTION
**Pull request by dorumd**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31135 
| License       | MIT
| Doc PR        | -

Prepare for PHP 7.4: use ReflectionReference in VarCloner

### CODE REVIEW

**Comment by dorumd** [comment]
@nicolas-grekas Ready for review.

File: src/Symfony/Component/VarDumper/Cloner/VarCloner.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Cloner/VarCloner.php]: I'd suggest keeping the comment here: it always applies here

File: src/Symfony/Component/VarDumper/Cloner/VarCloner.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Cloner/VarCloner.php]: should be removed

File: src/Symfony/Component/VarDumper/Tests/Dumper/CliDumperTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Tests/Dumper/CliDumperTest.php]: That's a PHP7.4 bug.
I reported it as https://bugs.php.net/77951
Please revert the test case and let's wait for the issue to be fixed before merging.

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Thanks, that's good step!
This highlights a PHP7.4 bug, reported as https://bugs.php.net/77951
Let's wait for the issue to be fixed before merging.

File: src/Symfony/Component/VarDumper/Cloner/VarCloner.php
Comment (linaori) [inline src/Symfony/Component/VarDumper/Cloner/VarCloner.php]: Can we rely on feature detection instead of version detection (and cache this result)? It seems that `class_exists(ReflectionReference::class)` would be an option.

Comment (nicolas-grekas) [reply src/Symfony/Component/VarDumper/Cloner/VarCloner.php]: Version detection allows opcache to optimize the generated opcodes.

**Comment by nicolas-grekas** [comment]
The bug has been fixed upstream, let's try again? Can you rerun the tests see if they pass now?

**Comment by nicolas-grekas** [comment]
Oh, would you mind doing a benchmark too? I'd be happy to know if the new way works faster that the current one.

**Comment by nicolas-grekas** [comment]
In theory, this should work, in practice, tests fail.
Waiting for feedback on https://bugs.php.net/78263
Could you please add a test case with a typed property that works by reference meanwhile?

File: src/Symfony/Component/VarDumper/Caster/Caster.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/Caster.php]: need to account for https://bugs.php.net/78319

File: src/Symfony/Component/VarDumper/Cloner/VarCloner.php
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Cloner/VarCloner.php]: On PHP 7.4, detecting a reference needs mutating the value when it is an array with refcount=1, for catching circular refs.
See https://github.com/php/php-src/pull/4380#issuecomment-513564327

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @dorumd.

**Pull request merged**
